damnit/backend/listener.py

```diff
@@ -54,38 +70,80 @@ def execute_direct(submitter, request):
     local_extraction_threads.append(extr)
     extr.start()
 
+class ListenerDB:
+    def __init__(self, db_dir):
+        self.conn = sqlite3.connect(db_dir.absolute() / "listener.sqlite")
+        self.conn.executescript(SCHEMA)
+        self._settings = KeyValueMapping(self.conn, "settings")
+
+        # To help prevent accidentally starting multiple listeners we default to
+        # putting it in static mode.
+        self.settings.setdefault("static_mode", True)
+        self.settings.setdefault("allow_local_processing", False)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.conn.close()
+
+    @property
+    def settings(self):
+        return self._settings
+
+    def all_proposals(self):
+        rows = self.conn.execute("SELECT proposal, db_dir, official FROM proposal_databases").fetchall()
+        result = { }
+        for proposal, db_dir, official in rows:
+            if proposal not in result:
+                result[proposal] = []
+            result[proposal].append(ProposalDBInfo(Path(db_dir), bool(official)))
```

Comment:
There's a shortcut for this - up to you if you think it's clearer or not:


```suggestion
            result.setdefault(proposal, []).append(
                ProposalDBInfo(Path(db_dir), bool(official))
            )
```